tokio-quiche/src/quic/scheduler.rs

```diff
@@ -0,0 +1,349 @@
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use itertools::Itertools;
+use rand::seq::SliceRandom;
+use rand::thread_rng;
+use std::fmt::Debug;
+use std::net::SocketAddr;
+use std::str::FromStr;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering::Relaxed;
+use std::sync::Arc;
+
+use crate::quic::QuicheConnection;
+
+pub type BoxedScheduler = Arc<dyn PacketScheduler + Send + Sync + 'static>;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[repr(C)]
+pub enum PacketSchedulingAlgorithm {
+    /// MinRTT Path Scheduler algorithm - selects path with lowest RTT
+    MinRTT        = 0,
+    /// Round Robin Path Scheduler algorithm - cycles through available paths
+    RoundRobin    = 1,
+    /// Random Path Scheduler algorithm - randomly selects among available paths
+    Random        = 2,
+    /// Lowest Latency Path Scheduler algorithm - like MinRTT but also considers
+    /// jitter
+    LowestLatency = 3,
+}
+
+impl PacketSchedulingAlgorithm {
+    /// Returns the name of the algorithm.
+    pub fn name(self) -> &'static str {
+        match self {
+            PacketSchedulingAlgorithm::MinRTT => "minrtt",
+            PacketSchedulingAlgorithm::RoundRobin => "roundrobin",
+            PacketSchedulingAlgorithm::Random => "random",
+            PacketSchedulingAlgorithm::LowestLatency => "lowestlatency",
+        }
+    }
+
+    fn active_paths(&self, conn: &QuicheConnection) -> Vec<quiche::PathStats> {
+        conn.path_stats()
+            .filter(|p| {
+                p.active && !matches!(p.state, quiche::PathState::Closed(_))
+            })
+            .collect()
+    }
+}
+
+impl FromStr for PacketSchedulingAlgorithm {
+    type Err = quiche::Error;
+
+    /// Converts a string to `PacketSchedulingAlgorithm`.
+    ///
+    /// If `name` is not valid, `Error::PacketScheduler` is returned.
+    fn from_str(name: &str) -> Result<Self, quiche::Error> {
+        match name {
+            "minrtt" => Ok(PacketSchedulingAlgorithm::MinRTT),
+            "roundrobin" => Ok(PacketSchedulingAlgorithm::RoundRobin),
+            "random" => Ok(PacketSchedulingAlgorithm::Random),
+            "lowestlatency" => Ok(PacketSchedulingAlgorithm::LowestLatency),
+            _ => Err(quiche::Error::PacketScheduler),
+        }
+    }
+}
+
+/// Trait defining a packet scheduler that can determine which path to use for
+/// sending packets
+pub trait PacketScheduler: Debug + Send + Sync + 'static {
+    /// Returns an ordered list of paths according to the scheduler's algorithm
+    fn schedule<'a>(
+        &self, conn: &'a QuicheConnection,
+    ) -> Box<dyn Iterator<Item = (SocketAddr, SocketAddr)> + 'a>;
```

Comment:
I am struggling a bit about this method. Unless I missed something, it seems unused in the whole code base. Furthermore, a call to the scheduler should not automatically perform memory allocations (as the returned `Box` type suggests).